File: hooks/scripts/_archived_ts/notifications/notify.ts
// Notification router — reads stdin JSON and dispatches to enabled providers (Telegram, Discord, Slack)
import { loadEnv } from "./lib/env-loader.ts";
import { telegram } from "./providers/telegram.ts";
import { discord } from "./providers/discord.ts";
import { slack } from "./providers/slack.ts";
import type { HookInput } from "./providers/telegram.ts";

const PROVIDERS = [telegram, discord, slack];

interface ProviderResult {
  provider: string;
  success: boolean;
  error?: string;
  throttled?: boolean;
}

async function readStdin(): Promise<HookInput> {
  if (process.stdin.isTTY) return {};

  return new Promise((resolve) => {
    let data = "";
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (chunk) => { data += chunk; });
    process.stdin.on("end", () => {
      if (!data.trim()) { resolve({}); return; }
      try { resolve(JSON.parse(data) as HookInput); }
      catch (err) {
        process.stderr.write(`[notify] Invalid JSON input: ${(err as Error).message}\n`);
        resolve({});
      }
    });
    process.stdin.on("error", (err) => {
      process.stderr.write(`[notify] Stdin error: ${err.message}\n`);
      resolve({});
    });
    setTimeout(() => { process.stderr.write("[notify] Stdin timeout\n"); resolve({}); }, 5000);
  });
}

function hasProviderEnv(prefix: string, env: Record<string, string>): boolean {
  return Object.keys(env).some((key) => key.startsWith(prefix + "_"));
}

async function main(): Promise<void> {
  try {
    const input = await readStdin();
    const cwd = input.cwd || process.cwd();
    const env = loadEnv(cwd);
    const results: ProviderResult[] = [];

    for (const provider of PROVIDERS) {
      const prefix = provider.name.toUpperCase();
      if (!hasProviderEnv(prefix, env)) continue;
      if (!provider.isEnabled(env)) continue;

      try {
        const result = await provider.send(input, env);
        results.push({ provider: provider.name, ...result });
        if (result.success) {
          process.stderr.write(`[notify] ${provider.name}: sent\n`);
        } else if (result.throttled) {
          process.stderr.write(`[notify] ${provider.name}: throttled\n`);
        } else {
          process.stderr.write(`[notify] ${provider.name}: failed - ${result.error}\n`);
        }
      } catch (err) {
        process.stderr.write(`[notify] ${provider.name} error: ${(err as Error).message}\n`);
        results.push({ provider: provider.name, success: false, error: (err as Error).message });
      }
    }

    if (results.length > 0) {
      const successful = results.filter((r) => r.success).length;
      process.stderr.write(`[notify] Summary: ${successful}/${results.length} succeeded\n`);
    }
  } catch (err) {
    process.stderr.write(`[notify] Fatal error: ${(err as Error).message}\n`);
  }

  process.exit(0); // Always exit 0 — never block Claude
}

main();

File: hooks/scripts/cmd/notify.go
// notify.go: notify subcommand — dispatches hook events to notification providers.
// Always exits 0 to avoid blocking Claude Code.
package cmd

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"
	"solon-hooks/internal/notify"
)

var notifyCmd = &cobra.Command{
	Use:   "notify",
	Short: "Handle Stop hook notifications",
	RunE:  runNotify,
}

func runNotify(cmd *cobra.Command, args []string) error {
	input := readNotifyStdin()

	cwd := ""
	if v, ok := input["cwd"]; ok {
		if s, ok := v.(string); ok {
			cwd = s
		}
	}
	if cwd == "" {
		cwd, _ = os.Getwd()
	}

	env := notify.LoadEnv(cwd)
	results := dispatchNotifications(input, env)

	if len(results) > 0 {
		successful := 0
		for _, r := range results {
			if r.success {
				successful++
			}
		}
		fmt.Fprintf(os.Stderr, "[notify] Summary: %d/%d succeeded\n", successful, len(results))
	}

	// Always exit 0 — never block Claude
	os.Exit(0)
	return nil
}

// notifyResult tracks per-provider send outcome for logging.
type notifyResult struct {
	provider  string
	success   bool
	errMsg    string
	throttled bool
}

// readNotifyStdin reads stdin with a 5-second timeout, parses JSON.
// Returns empty map on timeout, empty input, or parse error.
func readNotifyStdin() map[string]interface{} {
	ch := make(chan string, 1)
	go func() {
		data, _ := io.ReadAll(os.Stdin)
		ch <- string(data)
	}()

	var raw string
	select {
	case raw = <-ch:
	case <-time.After(5 * time.Second):
		fmt.Fprintln(os.Stderr, "[notify] Stdin timeout")
		return make(map[string]interface{})
	}

	raw = strings.TrimSpace(raw)
	if raw == "" {
		return make(map[string]interface{})
	}

	var result map[string]interface{}
	if err := json.Unmarshal([]byte(raw), &result); err != nil {
		fmt.Fprintf(os.Stderr, "[notify] Invalid JSON input: %s\n", err.Error())
		return make(map[string]interface{})
	}
	return result
}

// dispatchNotifications iterates all providers and sends if enabled.
func dispatchNotifications(input map[string]interface{}, env map[string]string) []notifyResult {
	var results []notifyResult

	for _, provider := range notify.Providers {
		prefix := strings.ToUpper(provider.Name) + "_"
		if !hasProviderEnv(prefix, env) {
			continue
		}
		if !provider.IsEnabled(env) {
			continue
		}

		res, err := provider.Send(input, env)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[notify] %s error: %s\n", provider.Name, err.Error())
			results = append(results, notifyResult{provider: provider.Name, errMsg: err.Error()})
			continue
		}

		if res.Success {
			fmt.Fprintf(os.Stderr, "[notify] %s: sent\n", provider.Name)
		} else if res.Throttled {
			fmt.Fprintf(os.Stderr, "[notify] %s: throttled\n", provider.Name)
		} else {
			fmt.Fprintf(os.Stderr, "[notify] %s: failed - %s\n", provider.Name, res.Error)
		}

		results = append(results, notifyResult{
			provider:  provider.Name,
			success:   res.Success,
			errMsg:    res.Error,
			throttled: res.Throttled,
		})
	}

	return results
}

// hasProviderEnv checks if any env key starts with the given prefix.
func hasProviderEnv(prefix string, env map[string]string) bool {
	for k := range env {
		if strings.HasPrefix(k, prefix) {
			return true
		}
	}
	return false
}
